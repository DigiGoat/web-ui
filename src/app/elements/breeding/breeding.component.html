<div class="row row-cols-1 row-cols-lg-3 color-scheme-tertiary py-3 my-5 justify-content-around" *ngIf="breeding">
  <div class="col"><app-goat-card [goat]="dam"></app-goat-card></div>
  <div class="col py-5 text-center">
    <div *ngIf="!breeding.kidded">
      <h3 *ngIf="breeding.exposed">Bred: {{breeding.exposed | longDate}}</h3>
      <br>
      <h3 *ngIf="breeding.due">Due: {{breeding.due | longDate}}</h3>
      <br>
      <!--<div *ngIf="breeding.does || breeding.wethers">
        <div *ngIf="breeding.does">
          <h5>Does: ${{breeding.does}}</h5>
        </div>
        <div *ngIf="breeding.wethers">
          <h5>Wethers: ${{breeding.wethers}}</h5>
        </div>
        <br>
      </div>
      <div *ngIf="breeding.reservations && breeding.reservations.length">
        <h5>Reservations:</h5>
        <div class="row justify-content-center">
          <div class="col" *ngFor="let reservation of breeding.reservations">{{reservation}}</div>
        </div>
      </div>
    </div>-->
    </div>
    <div *ngIf="breeding.kidded">
      <h3>Kidded: {{breeding.kidded | longDate}}</h3>
      <!--<h5>See the <a routerLink="/for-sale">For Sale</a> Page For More Info</h5>-->
    </div>
    <div [innerHTML]="breeding.description" markdown="breeding-{{breeding.dam}}x{{breeding.sire}}"></div>
  </div>
  <div class="col"><app-goat-card [goat]="sire"></app-goat-card></div>
</div>
<div class="row color-scheme-tertiary py-3 my-5 justify-content-around" *ngIf="!breeding">
  <div class="col"><app-goat-card></app-goat-card></div>
  <div class="col-auto pt-5">
    <div class="placeholder-wave">
      <h3 class="placeholder">Bred: Unknown</h3>
      <br>
      <h3 class="placeholder">Due: Unknown</h3>
      <br>
      <div>
        <h3 class="placeholder">Kidded: Unknown</h3>
      </div>
    </div>
  </div>
  <div class="col"><app-goat-card></app-goat-card></div>
</div>
