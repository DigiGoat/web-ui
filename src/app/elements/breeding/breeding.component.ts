import { Component, Input, type OnChanges } from '@angular/core';
import { findIDMatch, type Goat, type Kidding } from '../../services/goat/goat.service';

@Component({
  selector: 'app-breeding',
  templateUrl: './breeding.component.html',
  styleUrl: './breeding.component.scss',
  standalone: false
})
export class BreedingComponent implements OnChanges {
  ngOnChanges() {
    if (this.breeding) {
      this.dam = findIDMatch(this.breeding!.dam, [...(this.does ?? []), ...(this.references ?? [])]) ?? { name: this.breeding.dam };
      this.sire = findIDMatch(this.breeding!.sire, [...(this.bucks ?? []), ...(this.references ?? [])]) ?? { name: this.breeding.sire };
      //this.dam = this.does?.find(goat => goat.normalizeId === this.breeding!.dam) ?? this.references?.find(goat => goat.normalizeId === this.breeding!.dam) ?? { name: this.breeding.dam };
      //this.sire = this.bucks?.find(goat => goat.normalizeId === this.breeding!.sire) ?? this.references?.find(goat => goat.normalizeId === this.breeding!.sire) ?? { name: this.breeding.sire };
    }
  }
  @Input() breeding?: Kidding;
  @Input() does?: Goat[];
  @Input() bucks?: Goat[];
  @Input() references?: Goat[];

  dam?: Goat;
  sire?: Goat;

}
