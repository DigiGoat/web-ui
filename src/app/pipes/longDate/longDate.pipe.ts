import { DatePipe } from '@angular/common';
import { Pipe, PipeTransform } from '@angular/core';

@Pipe({
    name: 'longDate',
    standalone: false
})
export class LongDatePipe implements PipeTransform {

  transform(value?: string): string {
    if (!value || !Date.parse(value)) {
      return value ?? '';
    } else {
      const datePipe = new DatePipe('en-US');
      return datePipe.transform(value, 'longDate') ?? '';
    }
  }

}
