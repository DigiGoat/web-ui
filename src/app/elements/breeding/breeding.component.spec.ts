import { ComponentFixture, TestBed } from '@angular/core/testing';

import { LongDatePipe } from '../../pipes/longDate/longDate.pipe';
import { BreedingComponent } from './breeding.component';

describe('KiddingComponent', () => {
  let component: BreedingComponent;
  let fixture: ComponentFixture<BreedingComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [BreedingComponent, LongDatePipe]
    })
      .compileComponents();

    fixture = TestBed.createComponent(BreedingComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
  describe('With a Kidding (v1.0.0)', () => {
    beforeEach(() => {
      component.breeding = {
        dam: 'TEST_DAM',
        sire: 'TEST_SIRE',
        description: 'TEST_DETAILS',
        exposed: '1-2-34',
        due: '5-6-78',
      };
      component.does = [];
      component.bucks = [];
      component.ngOnChanges();
      fixture.detectChanges();
    });
    it('should match the snapshot', () => {
      expect(fixture).toMatchSnapshot();
    });
    it('should try to find the dam', () => {
      expect(component.dam).toBeDefined();
    });
    it('should try to find the sire', () => {
      expect(component.bucks).toBeDefined();
    });
    describe('With a previous Kidding', () => {
      beforeEach(() => {
        component.breeding = {
          ...component.breeding,
          kidded: '9-10-11',
        };
        component.ngOnChanges();
        fixture.detectChanges();
      });
      it('should match the snapshot', () => {
        expect(fixture).toMatchSnapshot();
      });
    });
  });
  describe('Without a Kidding', () => {
    beforeEach(() => {
      component.breeding = undefined;
      component.does = { find: jest.fn() } as any;
      component.bucks = { find: jest.fn() } as any;
      component.ngOnChanges();
      fixture.detectChanges();
    });
    it('should match the snapshot', () => {
      expect(fixture).toMatchSnapshot();
    });
    it('should NOT try to find the dam', () => {
      expect(component.does!.find).not.toHaveBeenCalled();
    });
    it('should NOT try to find the sire', () => {
      expect(component.bucks!.find).not.toHaveBeenCalled();
    });
  });
});
