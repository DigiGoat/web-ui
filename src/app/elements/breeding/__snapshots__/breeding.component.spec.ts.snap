// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KiddingComponent With a Kidding (v1.0.0) With a previous Kidding should match the snapshot 1`] = `
<app-breeding
  breeding={[Function Object]}
  bucks={[Function Array]}
  dam={[Function Object]}
  does={[Function Array]}
  sire={[Function Object]}
>
  <div
    class="row row-cols-1 row-cols-lg-3 color-scheme-tertiary py-3 my-5 justify-content-around"
  >
    <div
      class="col"
    >
      <app-goat-card />
    </div>
    <div
      class="col py-5 text-center"
    >
      <div>
        <h3>
          Kidded: September 10, 2011
        </h3>
      </div>
      <div>
        TEST_DETAILS
      </div>
    </div>
    <div
      class="col"
    >
      <app-goat-card />
    </div>
  </div>
</app-breeding>
`;

exports[`KiddingComponent With a Kidding (v1.0.0) should match the snapshot 1`] = `
<app-breeding
  breeding={[Function Object]}
  bucks={[Function Array]}
  dam={[Function Object]}
  does={[Function Array]}
  sire={[Function Object]}
>
  <div
    class="row row-cols-1 row-cols-lg-3 color-scheme-tertiary py-3 my-5 justify-content-around"
  >
    <div
      class="col"
    >
      <app-goat-card />
    </div>
    <div
      class="col py-5 text-center"
    >
      <div>
        <h3>
          Bred: January 2, 2034
        </h3>
        <br />
        <h3>
          Due: May 6, 1978
        </h3>
        <br />
      </div>
      <div>
        TEST_DETAILS
      </div>
    </div>
    <div
      class="col"
    >
      <app-goat-card />
    </div>
  </div>
</app-breeding>
`;

exports[`KiddingComponent Without a Kidding should match the snapshot 1`] = `
<app-breeding
  breeding="undefined"
  bucks={[Function Object]}
  does={[Function Object]}
>
  <div
    class="row color-scheme-tertiary py-3 my-5 justify-content-around"
  >
    <div
      class="col"
    >
      <app-goat-card />
    </div>
    <div
      class="col-auto pt-5"
    >
      <div
        class="placeholder-wave"
      >
        <h3
          class="placeholder"
        >
          Bred: Unknown
        </h3>
        <br />
        <h3
          class="placeholder"
        >
          Due: Unknown
        </h3>
        <br />
        <div>
          <h3
            class="placeholder"
          >
            Kidded: Unknown
          </h3>
        </div>
      </div>
    </div>
    <div
      class="col"
    >
      <app-goat-card />
    </div>
  </div>
</app-breeding>
`;
